# Imports

In [1]:

# Step 1: Imports & Setup

import pandas as pd
import numpy as np


# Loading data Into DataFrame 

In [2]:

# Step 2: Load the raw data file (.csv) into pandas dataframe

df = pd.read_csv("../data/data_raw/Loan_default.csv")


# Quick Initial Inspection of DataFrame

In [3]:

# Display a preview of the DataFrame
df


LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0       I38PQUQS96   56   85994       50587          520              80   
1       HPSK72WA7R   69   50432      124440          458              15   
2       C1OZ6DPJ8Y   46   84208      129188          451              26   
3       V2KKSFM3UN   32   31713       44799          743               0   
4       EY08JDHTZP   60   20437        9139          633               8   
...            ...  ...     ...         ...          ...             ...   
255342  8C6S86ESGC   19   37979      210682          541             109   
255343  98R4KDHNND   32   51953      189899          511              14   
255344  XQK1UUUNGP   56   84820      208294          597              70   
255345  JAO28CPL4H   42   85109       60575          809              40   
255346  ZTH91CGL0B   62   22418       18481          636             113   

        NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0                    4         15.23        36      0.44   Bachelor's   
1                    1          4.81        60      0.68     Master's   
2                    3         21.17        24      0.31     Master's   
3                    3          7.07        24      0.23  High School   
4                    4          6.51        48      0.73   Bachelor's   
...                ...           ...       ...       ...          ...   
255342               4         14.11        12      0.85   Bachelor's   
255343               2         11.55        24      0.21  High School   
255344               3          5.29        60      0.50  High School   
255345               1         20.90        48      0.44  High School   
255346               2          6.73        12      0.48   Bachelor's   

       EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0           Full-time      Divorced         Yes           Yes       Other   
1           Full-time       Married          No            No       Other   
2          Unemployed      Divorced         Yes           Yes        Auto   
3           Full-time       Married          No            No    Business   
4          Unemployed      Divorced          No           Yes        Auto   
...               ...           ...         ...           ...         ...   
255342      Full-time       Married          No            No       Other   
255343      Part-time      Divorced          No            No        Home   
255344  Self-employed       Married         Yes           Yes        Auto   
255345      Part-time        Single         Yes           Yes       Other   
255346     Unemployed      Divorced         Yes            No   Education   

       HasCoSigner  Default  
0              Yes        0  
1              Yes        0  
2               No        1  
3               No        0  
4               No        0  
...            ...      ...  
255342          No        0  
255343          No        1  
255344         Yes        0  
255345          No        0  
255346         Yes        0  

[255347 rows x 18 columns]

In [4]:

df.shape
# Returns a tuple (rows, columns) indicating the dimensions of the DataFrame 'df'


(255347, 18)


- `df.head(n)` : Shows the first 'n' rows of the DataFrame (default is 5)


In [5]:

df.head(10)


LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   
3  V2KKSFM3UN   32   31713       44799          743               0   
4  EY08JDHTZP   60   20437        9139          633               8   
5  A9S62RQ7US   25   90298       90448          720              18   
6  H8GXPAOS71   38  111188      177025          429              80   
7  0HGZQKJ36W   56  126802      155511          531              67   
8  1R0N3LGNRJ   36   42053       92357          827              83   
9  CM9L1GTT2P   40  132784      228510          480             114   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0               4         15.23        36      0.44   Bachelor's   
1               1          4.81        60      0.68     Master's   
2               3         21.17        24      0.31     Master's   
3               3          7.07        24      0.23  High School   
4               4          6.51        48      0.73   Bachelor's   
5               2         22.72        24      0.10  High School   
6               1         19.11        12      0.16   Bachelor's   
7               4          8.15        60      0.43          PhD   
8               1         23.94        48      0.20   Bachelor's   
9               4          9.09        48      0.33  High School   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   
3      Full-time       Married          No            No    Business   
4     Unemployed      Divorced          No           Yes        Auto   
5     Unemployed        Single         Yes            No    Business   
6     Unemployed        Single         Yes            No        Home   
7      Full-time       Married          No            No        Home   
8  Self-employed      Divorced         Yes            No   Education   
9  Self-employed       Married         Yes            No       Other   

  HasCoSigner  Default  
0         Yes        0  
1         Yes        0  
2          No        1  
3          No        0  
4          No        0  
5         Yes        1  
6         Yes        0  
7         Yes        0  
8          No        1  
9         Yes        0


- `df.tail(n)` : Shows the last 'n' rows of the DataFrame (default is 5)


In [6]:

df.tail(10)


LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
255337  DSL4O0KAWD   64   73743      140354          300               0   
255338  6V8S5IUS63   68   21711      168231          352              78   
255339  O6SWO6CBGB   51   69492      122962          348              66   
255340  48LOOK4VR1   41   61809      119238          444              34   
255341  AKXAXQN7PG   40  129890      116119          701              38   
255342  8C6S86ESGC   19   37979      210682          541             109   
255343  98R4KDHNND   32   51953      189899          511              14   
255344  XQK1UUUNGP   56   84820      208294          597              70   
255345  JAO28CPL4H   42   85109       60575          809              40   
255346  ZTH91CGL0B   62   22418       18481          636             113   

        NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
255337               2          4.12        12      0.24          PhD   
255338               2          9.71        60      0.36          PhD   
255339               2         10.83        48      0.27  High School   
255340               2         19.99        36      0.31     Master's   
255341               3          9.91        24      0.23  High School   
255342               4         14.11        12      0.85   Bachelor's   
255343               2         11.55        24      0.21  High School   
255344               3          5.29        60      0.50  High School   
255345               1         20.90        48      0.44  High School   
255346               2          6.73        12      0.48   Bachelor's   

       EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
255337  Self-employed        Single         Yes            No   Education   
255338      Full-time      Divorced         Yes           Yes        Home   
255339      Part-time      Divorced          No            No        Home   
255340      Part-time       Married         Yes           Yes        Auto   
255341      Part-time      Divorced         Yes            No        Home   
255342      Full-time       Married          No            No       Other   
255343      Part-time      Divorced          No            No        Home   
255344  Self-employed       Married         Yes           Yes        Auto   
255345      Part-time        Single         Yes           Yes       Other   
255346     Unemployed      Divorced         Yes            No   Education   

       HasCoSigner  Default  
255337         Yes        0  
255338          No        0  
255339          No        0  
255340         Yes        0  
255341         Yes        1  
255342          No        0  
255343          No        1  
255344         Yes        0  
255345          No        0  
255346         Yes        0


### `df.sample()` :

- In this method, a parameter to specify the size of the sample is mandatory.

- You must provide either 'n' or 'frac', but not both :

    - `n` : Specifies the number of rows to return in the sample. For example, `df.sample(n=10)` will return exactly 10 random rows.

    - `frac` : Specifies the fraction of the total rows to return. For example, `df.sample(frac=0.5)` will return 50% of the DataFrame's rows.


In [7]:

df.sample(n = 7)   # will return exactly 7 random rows from the dataframe 'df'


LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
4426    EG4UTGOQSP   37   31962      217953          442              11   
92438   JMD9L04YA8   42  127444      172696          718             112   
29191   FCTGVGF3XT   45   83962       23858          714              69   
46785   IAXMWHGQTI   36   89077      212019          482              98   
214151  M0LCC4XXYP   34   93250      109372          388              84   
154861  Q5RAYO6LE2   25   42478      247323          539              69   
173435  SBF3XZRJ4D   48   23119      144320          329              19   

        NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
4426                 3          7.59        48      0.11   Bachelor's   
92438                3          3.05        48      0.15     Master's   
29191                2          3.49        36      0.35   Bachelor's   
46785                4         13.26        48      0.49   Bachelor's   
214151               1          5.85        36      0.84     Master's   
154861               1         11.89        60      0.78  High School   
173435               3          7.24        12      0.18   Bachelor's   

       EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
4426        Part-time      Divorced         Yes            No        Home   
92438      Unemployed       Married          No            No   Education   
29191   Self-employed       Married         Yes           Yes   Education   
46785       Part-time       Married         Yes           Yes        Auto   
214151      Full-time       Married         Yes           Yes    Business   
154861  Self-employed        Single         Yes            No    Business   
173435      Full-time      Divorced          No           Yes    Business   

       HasCoSigner  Default  
4426           Yes        1  
92438           No        0  
29191           No        0  
46785           No        0  
214151          No        0  
154861         Yes        1  
173435         Yes        0

In [8]:

df.sample(frac = 0.001)    # will return (0.001 * 100) % random rows from the dataframe 'df'


LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
110571  BQP7I5ELEW   43   26062      233034          640              59   
86531   8ZFL0Z92O1   29  123713      204589          632              79   
203131  13QLPST5D8   43   77568      116712          316             101   
76454   CGFNCYAIEO   54   18794      145570          511              24   
180416  KH98FE1WPB   64   55185       81316          719              46   
...            ...  ...     ...         ...          ...             ...   
106609  OK0WKQ4L0Q   18   76262       47498          302              63   
171994  BQWQXXTHSU   67   64257      130672          348              76   
14408   KMOD47D6VJ   42   26435      100546          685              71   
65701   W27SI6708U   69  104354      195431          425              59   
225566  DYIQ8VDQPW   65   97525       83640          492              51   

        NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
110571               3         11.12        36      0.88  High School   
86531                4          5.01        24      0.48   Bachelor's   
203131               3         13.37        36      0.53  High School   
76454                4         23.38        36      0.36  High School   
180416               2          7.02        36      0.71  High School   
...                ...           ...       ...       ...          ...   
106609               1          8.41        24      0.71          PhD   
171994               3          3.02        12      0.47   Bachelor's   
14408                1         10.96        48      0.14     Master's   
65701                2          3.82        36      0.60          PhD   
225566               4         19.72        60      0.59     Master's   

       EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
110571      Full-time       Married          No           Yes        Home   
86531   Self-employed      Divorced         Yes           Yes        Home   
203131  Self-employed      Divorced          No           Yes        Home   
76454   Self-employed        Single         Yes           Yes   Education   
180416      Full-time      Divorced          No           Yes        Auto   
...               ...           ...         ...           ...         ...   
106609      Part-time       Married          No            No       Other   
171994     Unemployed       Married         Yes            No        Auto   
14408   Self-employed        Single          No            No        Auto   
65701      Unemployed        Single          No            No       Other   
225566      Part-time        Single          No           Yes        Home   

       HasCoSigner  Default  
110571         Yes        0  
86531          Yes        0  
203131          No        0  
76454           No        1  
180416          No        0  
...            ...      ...  
106609          No        0  
171994          No        0  
14408          Yes        0  
65701           No        0  
225566         Yes        0  

[255 rows x 18 columns]

In [9]:

# Provides a concise summary of the DataFrame 'df', including the data types, non-null values, and memory usage
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255347 entries, 0 to 255346
Data columns (total 18 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   LoanID          255347 non-null  object 
 1   Age             255347 non-null  int64  
 2   Income          255347 non-null  int64  
 3   LoanAmount      255347 non-null  int64  
 4   CreditScore     255347 non-null  int64  
 5   MonthsEmployed  255347 non-null  int64  
 6   NumCreditLines  255347 non-null  int64  
 7   InterestRate    255347 non-null  float64
 8   LoanTerm        255347 non-null  int64  
 9   DTIRatio        255347 non-null  float64
 10  Education       255347 non-null  object 
 11  EmploymentType  255347 non-null  object 
 12  MaritalStatus   255347 non-null  object 
 13  HasMortgage     255347 non-null  object 
 14  HasDependents   255347 non-null  object 
 15  LoanPurpose     255347 non-null  object 
 16  HasCoSigner     255347 non-null  object 
 17  Default   

In [10]:

# Generates descriptive statistics (like mean, std, min, max, etc.) for 'numerical columns' only
df.describe()


Age         Income     LoanAmount    CreditScore  \
count  255347.000000  255347.000000  255347.000000  255347.000000   
mean       43.498306   82499.304597  127578.865512     574.264346   
std        14.990258   38963.013729   70840.706142     158.903867   
min        18.000000   15000.000000    5000.000000     300.000000   
25%        31.000000   48825.500000   66156.000000     437.000000   
50%        43.000000   82466.000000  127556.000000     574.000000   
75%        56.000000  116219.000000  188985.000000     712.000000   
max        69.000000  149999.000000  249999.000000     849.000000   

       MonthsEmployed  NumCreditLines   InterestRate       LoanTerm  \
count   255347.000000   255347.000000  255347.000000  255347.000000   
mean        59.541976        2.501036      13.492773      36.025894   
std         34.643376        1.117018       6.636443      16.969330   
min          0.000000        1.000000       2.000000      12.000000   
25%         30.000000        2.000000       7.770000      24.000000   
50%         60.000000        2.000000      13.460000      36.000000   
75%         90.000000        3.000000      19.250000      48.000000   
max        119.000000        4.000000      25.000000      60.000000   

            DTIRatio        Default  
count  255347.000000  255347.000000  
mean        0.500212       0.116128  
std         0.230917       0.320379  
min         0.100000       0.000000  
25%         0.300000       0.000000  
50%         0.500000       0.000000  
75%         0.700000       0.000000  
max         0.900000       1.000000

In [11]:

# '.dtypes' returns a Pandas Series with the data type of each column
result = df.dtypes
print(f"\nResult :\n\n{result}")      # Printing the 'result'
print(f"\nThe type of above 'Result' : {type(result)}")   # Printing the type of 'result'



Result :

LoanID             object
Age                 int64
Income              int64
LoanAmount          int64
CreditScore         int64
MonthsEmployed      int64
NumCreditLines      int64
InterestRate      float64
LoanTerm            int64
DTIRatio          float64
Education          object
EmploymentType     object
MaritalStatus      object
HasMortgage        object
HasDependents      object
LoanPurpose        object
HasCoSigner        object
Default             int64
dtype: object

The type of above 'Result' : <class 'pandas.core.series.Series'>


In [12]:

# Provides an index of all column names
df.columns


Index(['LoanID', 'Age', 'Income', 'LoanAmount', 'CreditScore',
       'MonthsEmployed', 'NumCreditLines', 'InterestRate', 'LoanTerm',
       'DTIRatio', 'Education', 'EmploymentType', 'MaritalStatus',
       'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner',
       'Default'],
      dtype='object')

In [13]:

# Provides information about the row index
df.index


RangeIndex(start=0, stop=255347, step=1)

In [14]:

# Counts the number of missing values in each column
df.isnull().sum()


LoanID            0
Age               0
Income            0
LoanAmount        0
CreditScore       0
MonthsEmployed    0
NumCreditLines    0
InterestRate      0
LoanTerm          0
DTIRatio          0
Education         0
EmploymentType    0
MaritalStatus     0
HasMortgage       0
HasDependents     0
LoanPurpose       0
HasCoSigner       0
Default           0
dtype: int64

In [15]:

# Counts the total number of duplicate rows
df.duplicated().sum()


np.int64(0)

In [16]:

# Counts the number of unique values in each column
df.nunique()


LoanID            255347
Age                   52
Income            114620
LoanAmount        158729
CreditScore          550
MonthsEmployed       120
NumCreditLines         4
InterestRate        2301
LoanTerm               5
DTIRatio              81
Education              4
EmploymentType         4
MaritalStatus          3
HasMortgage            2
HasDependents          2
LoanPurpose            5
HasCoSigner            2
Default                2
dtype: int64

In [17]:

# List of Available Columns
df.columns.tolist()


['LoanID',
 'Age',
 'Income',
 'LoanAmount',
 'CreditScore',
 'MonthsEmployed',
 'NumCreditLines',
 'InterestRate',
 'LoanTerm',
 'DTIRatio',
 'Education',
 'EmploymentType',
 'MaritalStatus',
 'HasMortgage',
 'HasDependents',
 'LoanPurpose',
 'HasCoSigner',
 'Default']

In [18]:

# Missing Values
df.isna().sum().sort_values(ascending=False).head(20)


LoanID            0
Age               0
Income            0
LoanAmount        0
CreditScore       0
MonthsEmployed    0
NumCreditLines    0
InterestRate      0
LoanTerm          0
DTIRatio          0
Education         0
EmploymentType    0
MaritalStatus     0
HasMortgage       0
HasDependents     0
LoanPurpose       0
HasCoSigner       0
Default           0
dtype: int64

In [19]:

# Unique Values (Cardinality)
df.nunique().sort_values(ascending=False).head(20)  # top 20


LoanID            255347
LoanAmount        158729
Income            114620
InterestRate        2301
CreditScore          550
MonthsEmployed       120
DTIRatio              81
Age                   52
LoanTerm               5
LoanPurpose            5
NumCreditLines         4
Education              4
EmploymentType         4
MaritalStatus          3
HasMortgage            2
HasDependents          2
HasCoSigner            2
Default                2
dtype: int64

In [20]:

# Target Column Check
if "loan_status" in df.columns:
    print(df["loan_status"].value_counts())
elif "repay_fail" in df.columns:
    print(df["repay_fail"].value_counts())
elif "Default" in df.columns:
    print(df["Default"].value_counts())
else:
    print("⚠️ Could not find target column yet.")


Default
0    225694
1     29653
Name: count, dtype: int64


In [21]:

# Target Balance (Normalized)

# Check class balance of the target column (in percentages)
if "Default" in df.columns:
    print(df["Default"].value_counts(normalize=True))
else:
    print("⚠️ Could not find target column 'Default'")


Default
0    0.883872
1    0.116128
Name: proportion, dtype: float64


In [22]:

# Group columns by type

numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = df.select_dtypes(include=["object"]).columns.tolist()

print("\nNumeric columns:", numeric_cols[:10], "\n\ntotal:", len(numeric_cols))
print("\nCategorical columns:", categorical_cols[:10], "\n\ntotal:", len(categorical_cols),"\n")



Numeric columns: ['Age', 'Income', 'LoanAmount', 'CreditScore', 'MonthsEmployed', 'NumCreditLines', 'InterestRate', 'LoanTerm', 'DTIRatio', 'Default'] 

total: 10

Categorical columns: ['LoanID', 'Education', 'EmploymentType', 'MaritalStatus', 'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner'] 

total: 8 



## 🔍 Quick Observations

- **Shape:** `2,55,347 rows × 18 columns`  

- **Target Column:** `Default`

- **Class Balance:**  
  - 0 → Majority (Non-defaults)  
  - 1 → Minority (Defaults)  
  - `Default
    0  :  0.883872 %
    1  :  0.116128 %`

- **Missing Values (Top Columns):**  
  - No such columns

- **Example Numerical Columns:**  
  - `Age`, `Income`, `LoanAmount`, `CreditScore`, `MonthsEmployed`, `NumCreditLines`, `InterestRate`, `LoanTerm`, `DTIRatio`, `Default` 

  - total: 10

- **Example Categorical Columns:**  
  - `LoanID`, `Education`, `EmploymentType`, `MaritalStatus`, `HasMortgage`, `HasDependents`, `LoanPurpose`, `HasCoSigner`
  -  total: 8

- **Suspicious Columns to Drop (ID-like or non-informative):**  
  - `LoanID` (unique identifier, no predictive power)

---
